### PR TITLE
Update hover before other mouse event propagates

### LINKED ISSUE #3153: Hovering should be updated before any mouse event propagations
Hovering should be updated before any mouse event propagations to avoid incorrect `IsHovered` values inside event as can be seen in the test.

```csharp
    public class TestSceneHoverUpdates : ManualInputManagerTestScene
    {
        private TestDrawable hoveredDrawable;

        [SetUp]
        public override void SetUp() => Schedule(() =>
        {
            base.SetUp();

            Child = hoveredDrawable = new TestDrawable();
        });

        [Test]
        public void TestCheckHoverFromMouseMove()
        {
            AddStep("move mouse inside drawable", () => InputManager.MoveMouseTo(hoveredDrawable));
            AddAssert("is hover correct", () => !hoveredDrawable.IncorrectHover);
        }

        private class TestDrawable : Box
        {
            public bool IncorrectHover = false;

            public TestDrawable()
            {
                Anchor = Anchor.Centre;
                Origin = Anchor.Centre;
                Size = new Vector2(128);
            }

            protected override bool OnMouseMove(MouseMoveEvent e)
            {
                if (!IncorrectHover && IsHovered != ReceivePositionalInputAt(e.ScreenSpaceMousePosition))
                    IncorrectHover = true;

                return base.OnMouseMove(e);
            }
        }
    }
```

### PR DESCRIPTION
**Pull request by mcendu**
Fixes #3153

### CODE REVIEW

File: osu.Framework.Tests/Visual/Input/TestSceneHoverUpdates.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+        private TestDrawable hoveredDrawable;
+
+        [SetUp]
+        public override void SetUp() => Schedule(() =>
+        {
+            base.SetUp();
+
+            Child = hoveredDrawable = new TestDrawable();
+
+            InputManager.MoveMouseTo(Vector2.Zero);
```
Comment (iiSaLMaN) [inline osu.Framework.Tests/Visual/Input/TestSceneHoverUpdates.cs]: you can just override `InitialMousePosition` to `Vector2.Zero`

**Comment by smoogipoo** [comment]
This is changing the ordering of events. You've changed:

```
1 -> OnMouseMove
2 -> OnHover
```

into

```
1 -> OnHover
2 -> OnMouseMove
```

This can't happen. As for the underlying issue, I'd need more time to convince myself that `IsHovered` should be set before any movement event because I'm not sure that's correct.
Is there a scenario which you are running into that's dependent on this being resolved? @iiSaLMaN

**Comment by iiSaLMaN** [comment]
> Is there a scenario which you are running into that's dependent on this being resolved? @iiSaLMaN

yes, the osu! ruleset seems to be entirely broken with my touch input implementation due to tapping in a circle while the hover is not updated in here: https://github.com/ppy/osu/blob/master/osu.Game.Rulesets.Osu/Objects/Drawables/DrawableHitCircle.cs#L196, `OnPressed(action)` called from a `MouseDownEvent`.

**Comment by smoogipoo** [comment]
That sounds like you should separate your `IInput` into a position and a button change. Mouse handlers already do the same thing.

**Comment by iiSaLMaN** [comment]
I already have `TouchButtonInput` and `TouchPositionInput`, but there is a case where `TouchButtonInput` can be applied with a pointer at position X:
`new TouchButtonInput(new TouchPointer(source, pos), true)`

**Comment by iiSaLMaN** [comment]
@smoogipoo If you meant separating the input in a way that a `TouchPositionInput` updates the hover then `TouchButtonInput` presses the object, then i'm not sure how that'll work unless you meant to make `TouchPositionInput` act completely as a `MousePositionAbsoluteInput`, then that would be too complicated to make work.

I suggest that all the events that are to be propagated gets saved up in a list, and after the manager finishes from applying handler's pending inputs, the events will be propagated starting from the most important (mouse event, key event, or joystick event) to the least important. (drag event, hover event and focus event), sounds fair?

**Comment by iiSaLMaN** [comment]
After trying to implement events sorting by most important to least important, I've found out that `MouseButtonEventManager` is too complicated to implement in, will see what i can do with your suggestion.

**Comment by iiSaLMaN** [comment]
@mcendu Are you planning on resolving conversations and fixing the event order?

**Pull request closed without merging**